### PR TITLE
feat: HyperlaneDb trait

### PR DESCRIPTION
**Pull request by daniel-savu**
### Description

Defines a `HyperlaneDb` trait for use in the validator submitter, so a mock can be passed in to simulate indexed data from reorged state.

Although the `HyperlaneDb` trait can't be `Clone` to be object safe, as long as we use `Arc`s everywhere, all will be ok. Eventually the aim is to replace all direct usage of `HyperlaneRocksDb` with `Arc<dyn HyperlaneDb>`.

### CODE REVIEW

**Comment by ameten** [comment]
Is the aim eventually to make `HyperlaneSqlDb` to implement `HyperlaneDb` trait?

File: rust/main/hyperlane-base/src/db/error.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /// Error opening the database
+    OpeningError {
+        /// Rocksdb error during opening
+        #[source]
+        source: rocksdb::Error,
+        /// Raw database path provided
+        path: PathBuf,
+        /// Parsed path used
+        canonicalized: PathBuf,
+    },
```
Comment (ameten) [inline rust/main/hyperlane-base/src/db/error.rs:21]: I wonder if we should hide the error from a particular type of database and expose a generalised error to the user of `HyperlaneDb` trait?

**Review by ameten — APPROVED** [review]

File: rust/main/hyperlane-base/src/db/storage_types.rs [hunk 2]
```diff
@@ -8,15 +8,18 @@ use hyperlane_core::{
...
+pub struct InterchainGasPaymentData {
+    /// The amount of tokens paid for the gas.
     pub payment: U256,
+    /// The amount of gas paid for.
     pub gas_amount: U256,
 }
 
 /// Subset of `InterchainGasExpenditure` excluding the message id which is
 /// stored in the key.
+#[allow(missing_docs)]
```
Comment (tkporter) [inline rust/main/hyperlane-base/src/db/storage_types.rs:20]: just wanna make sure this is intentional as it seems like some docs were added

**Review by tkporter — APPROVED** [review]

File: rust/main/config/testnet_config.json [hunk 3]
```diff
@@ -1446,6 +1446,194 @@
       "index": {
         "from": 1921514
       }
+    },
+    "test1": {
```
[the remark this answers is not in the document]
Comment (daniel-savu) [reply rust/main/config/testnet_config.json:1450]: they're ephemeral but deterministic, so thought it's reasonable to add them

File: rust/main/hyperlane-base/src/db/storage_types.rs (hunk 2 above)
Comment (daniel-savu) [reply rust/main/hyperlane-base/src/db/storage_types.rs:20]: I wasn't fully sure what the fields mean here tbh and I preferred under-documenting that mis-documenting

File: rust/main/hyperlane-base/src/db/error.rs (hunk 1 above)
Comment (daniel-savu) [reply rust/main/hyperlane-base/src/db/error.rs:21]: `DbError` is generically named, and its members could be anything based on specific implementations - `DbError::RocksError` can only be created from a rocksdb error so it won't be misused. 

`DbError` _is_ made fully generalised when it's converted to `ChainCommunicationError`, and I considered defining the error as an associated type, but usage ends up looking _more_ concrete, e.g. `Arc<dyn HyperlaneDb<DbError = RocksDbError>>`

Due to that I'm not sure it's worth generalising further in the current state, could you give an example?

Comment (ameten) [reply rust/main/hyperlane-base/src/db/error.rs:21]: If we just converting into a different error, we don't need to generalise. If we would like to react on the error in a meaningful way, we may want to generalise it.

**Pull request merged**